spec/unit/resolver_spec.rb

```diff
@@ -550,6 +510,152 @@ def resolve
+        sources = SourcesManager.sources(%w(master test_repo))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        spec = resolver.resolve.values.flatten.first
+        spec.version.to_s.should == '1.4'
+        spec.defined_in_file.should == fixture('spec-repos/master/Specs/JSONKit/1.4/JSONKit.podspec.json')
+
+        sources = SourcesManager.sources(%w(test_repo master))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        spec = resolver.resolve.values.flatten.first
+        spec.version.to_s.should == '1.4'
+        resolver.resolve.values.flatten.first.defined_in_file.should == fixture('spec-repos/test_repo/JSONKit/1.4/JSONKit.podspec')
+
+        UI.warnings.should.match /multiple specifications/
+      end
+
+      it 'does not warn when multiple sources contain a pod but a dependency ' \
+         'has an explicit source specified' do
+        test_repo_url = SourcesManager.source_with_name_or_url('test_repo').url
+        podfile = Podfile.new do
+          platform :ios
+          pod 'JSONKit', '1.4', :source => test_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(master test_repo))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        resolver.resolve
+
+        UI.warnings.should.not.match /multiple specifications/
+      end
+
+      it 'fails to resolve a dependency with an explicit source even if it can be ' \
+         'resolved using the global sources' do
+        test_repo_url = SourcesManager.source_with_name_or_url('test_repo').url
+        podfile = Podfile.new do
+          platform :ios
+          pod 'JSONKit', '1.5pre', :source => test_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(master))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        e = lambda { resolver.resolve }.should.raise Informative
+        e.message.should.match(/None of the spec sources contain a spec satisfying/)
+        e.message.should.match(/JSONKit/)
+        e.message.should.match(/\= 1.5pre/)
+      end
+
+      it 'resolves a dependency with an explicit source even if it can\'t be ' \
+         'resolved using the global sources' do
+        master_repo_url = SourcesManager.source_with_name_or_url('master').url
+        podfile = Podfile.new do
+          platform :ios
+          pod 'JSONKit', '1.5pre', :source => master_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(test_repo))
+        sources.map(&:url).should.not.include(master_repo_url)
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        spec = resolver.resolve.values.flatten.first
+        spec.version.to_s.should == '1.5pre'
+        spec.defined_in_file.should == fixture('spec-repos/master/Specs/JSONKit/1.5pre/JSONKit.podspec.json')
+      end
+
+      it 'uses explicit source repos for a dependency even when it\'s transitive' do
```

Comment:
Tests that if a dependency has transitive dependencies, and those transitive dependencies are also top-level dependencies, then the transitive dependency is resolved using the same resolution as the top-level dependency.